### ISSUE TITLE
[SuperEditor] - Document layout access error due to edit timing

### ISSUE THREAD

**Issue by matthew-carroll** [superlistapp/super_editor#1330]
**Update:**
After digging into this further, the problem is not specific to tagging. It's an issue with edit change notifications, which pieces of the tree call `setState()`, and when. This is a nuanced complication within `ContentLayer`s.

**Original Description:**
I noticed in the Example app that we're now getting errors when typing into the tagging demos. The issue seems limited to the tagging demos. We must have changed something recently that broke something in the tagging order of operations.

The technical problem is that Super Text Layout hasn't run layout when we access it, but I'm not sure why this problem just showed up.

Repro: open a tagging demo in the Example app, select the editor, start typing. Errors are repeatedly logged to the console.

```
flutter: Initializing logger: ExampleApp
flutter: _updateActionTagList()
flutter: _updateActionTagList()

======== Exception caught by widgets library =======================================================
The following _TypeError was thrown building SelectionLeadersDocumentLayer(dirty, state: _SelectionLeadersDocumentLayerState#2a17e):
Null check operator used on a null value

The relevant error-causing widget was: 
  SelectionLeadersDocumentLayer SelectionLeadersDocumentLayer:file:///Users/matt/Projects/super_editor/super_editor/lib/src/default_editor/super_editor.dart:670:12
When the exception was thrown, this was the stack: 
#0      SuperTextState.textLayout (package:super_text_layout/src/super_text.dart:70:89)
#1      _SuperTextWithSelectionState.textLayout (package:super_text_layout/src/super_text_layout_with_selection.dart:97:85)
#2      TextComponentState.textLayout (package:super_editor/src/default_editor/text.dart:459:60)
#3      TextComponentState.getOffsetForPosition (package:super_editor/src/default_editor/text.dart:482:12)
#4      TextComponentState.getRectForPosition (package:super_editor/src/default_editor/text.dart:491:20)
#5      _SingleColumnDocumentLayoutState.getRectForPosition (package:super_editor/src/default_editor/layout_single_column/_layout.dart:232:37)
#6      _SelectionLeadersDocumentLayerState._positionCaret (package:super_editor/src/infrastructure/selection_leader_document_layer.dart:112:31)
#7      _SelectionLeadersDocumentLayerState.build (package:super_editor/src/infrastructure/selection_leader_document_layer.dart:129:5)
#8      StatefulElement.build (package:flutter/src/widgets/framework.dart:5569:27)
#9      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5457:15)
#10     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5620:11)
#11     Element.rebuild (package:flutter/src/widgets/framework.dart:5173:7)
#12     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2881:19)
#13     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:957:21)
#14     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:457:5)
#15     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1325:15)
#16     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1255:9)
#17     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1113:5)
#18     _invoke (dart:ui/hooks.dart:170:13)
#19     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:410:5)
#20     _drawFrame (dart:ui/hooks.dart:140:31)
====================================================================================================

======== Exception caught by widgets library =======================================================
The following _TypeError was thrown building CaretDocumentOverlay(dirty, state: _CaretDocumentOverlayState#bf838(ticker active)):
Null check operator used on a null value

The relevant error-causing widget was: 
  CaretDocumentOverlay CaretDocumentOverlay:file:///Users/matt/Projects/super_editor/super_editor/lib/src/default_editor/super_editor.dart:837:14
When the exception was thrown, this was the stack: 
#0      SuperTextState.textLayout (package:super_text_layout/src/super_text.dart:70:89)
#1      _SuperTextWithSelectionState.textLayout (package:super_text_layout/src/super_text_layout_with_selection.dart:97:85)
#2      TextComponentState.textLayout (package:super_editor/src/default_editor/text.dart:459:60)
#3      TextComponentState.getOffsetForPosition (package:super_editor/src/default_editor/text.dart:482:12)
#4      TextComponentState.getRectForPosition (package:super_editor/src/default_editor/text.dart:491:20)
#5      _SingleColumnDocumentLayoutState.getRectForPosition (package:super_editor/src/default_editor/layout_single_column/_layout.dart:232:37)
#6      _CaretDocumentOverlayState._positionCaret (package:super_editor/src/default_editor/document_caret_overlay.dart:130:29)
#7      _CaretDocumentOverlayState.build (package:super_editor/src/default_editor/document_caret_overlay.dart:135:5)
#8      StatefulElement.build (package:flutter/src/widgets/framework.dart:5569:27)
#9      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5457:15)
#10     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5620:11)
#11     Element.rebuild (package:flutter/src/widgets/framework.dart:5173:7)
#12     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2881:19)
#13     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:957:21)
#14     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:457:5)
#15     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1325:15)
#16     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1255:9)
#17     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1113:5)
#18     _invoke (dart:ui/hooks.dart:170:13)
#19     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:410:5)
#20     _drawFrame (dart:ui/hooks.dart:140:31)
====================================================================================================

======== Exception caught by widgets library =======================================================
The following _TypeError was thrown building SelectionLeadersDocumentLayer(dirty, state: _SelectionLeadersDocumentLayerState#2a17e):
Null check operator used on a null value

The relevant error-causing widget was: 
  SelectionLeadersDocumentLayer SelectionLeadersDocumentLayer:file:///Users/matt/Projects/super_editor/super_editor/lib/src/default_editor/super_editor.dart:670:12
When the exception was thrown, this was the stack: 
#0      SuperTextState.textLayout (package:super_text_layout/src/super_text.dart:70:89)
#1      _SuperTextWithSelectionState.textLayout (package:super_text_layout/src/super_text_layout_with_selection.dart:97:85)
#2      TextComponentState.textLayout (package:super_editor/src/default_editor/text.dart:459:60)
#3      TextComponentState.getOffsetForPosition (package:super_editor/src/default_editor/text.dart:482:12)
#4      TextComponentState.getRectForPosition (package:super_editor/src/default_editor/text.dart:491:20)
#5      _SingleColumnDocumentLayoutState.getRectForPosition (package:super_editor/src/default_editor/layout_single_column/_layout.dart:232:37)
#6      _SelectionLeadersDocumentLayerState._positionCaret (package:super_editor/src/infrastructure/selection_leader_document_layer.dart:112:31)
#7      _SelectionLeadersDocumentLayerState.build (package:super_editor/src/infrastructure/selection_leader_document_layer.dart:129:5)
#8      StatefulElement.build (package:flutter/src/widgets/framework.dart:5569:27)
#9      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5457:15)
#10     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5620:11)
#11     Element.rebuild (package:flutter/src/widgets/framework.dart:5173:7)
#12     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2881:19)
#13     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:957:21)
#14     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:457:5)
#15     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1325:15)
#16     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1255:9)
#17     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1113:5)
#18     _invoke (dart:ui/hooks.dart:170:13)
#19     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:410:5)
#20     _drawFrame (dart:ui/hooks.dart:140:31)
====================================================================================================
```

**Comment by Jethro87** [comment]
I just tried to update super_editor to stable from the pegged version I have been using (before the atomic commands merge) and ran into these errors. I followed `example_editor.dart` to upgrade my code, and haven't done any tagging implementation as of yet. I'm targeting `stable` for all super_editor dependencies. 

cc @matthew-carroll 

```
═══════ Exception caught by widgets library ═══════════════════════════════════
The following _TypeError was thrown building SelectionLeadersDocumentLayer(dirty, state: _SelectionLeadersDocumentLayerState#4689a):
Null check operator used on a null value

The relevant error-causing widget was
SuperEditor-[<712833407>]
editor.dart:1150
When the exception was thrown, this was the stack
#0      SuperTextState.textLayout
super_text.dart:73
#1      _SuperTextWithSelectionState.textLayout
super_text_layout_with_selection.dart:100
#2      TextComponentState.textLayout
text.dart:462
#3      TextComponentState.getOffsetForPosition
text.dart:485
#4      TextComponentState.getRectForPosition
text.dart:494
#5      _SingleColumnDocumentLayoutState.getRectForPosition
_layout.dart:232
#6      _SelectionLeadersDocumentLayerState._positionCaret
selection_leader_document_layer.dart:112
#7      _SelectionLeadersDocumentLayerState.build
selection_leader_document_layer.dart:129
#8      StatefulElement.build
framework.dart:5409
#9      ComponentElement.performRebuild
framework.dart:5297
#10     StatefulElement.performRebuild
framework.dart:5462
#11     Element.rebuild
framework.dart:5016
#12     BuildOwner.buildScope
framework.dart:2779
#13     WidgetsBinding.drawFrame
binding.dart:916
#14     RendererBinding._handlePersistentFrameCallback
binding.dart:360
#15     SchedulerBinding._invokeFrameCallback
binding.dart:1297
#16     SchedulerBinding.handleDrawFrame
binding.dart:1227
#17     SchedulerBinding._handleDrawFrame
binding.dart:1085
#18     _invoke (dart:ui/hooks.dart:170:13)
#19     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:401:5)
#20     _drawFrame (dart:ui/hooks.dart:140:31)
════════════════════════════════════════════════════════════════════════════════
```

**Comment by matthew-carroll** [comment]
Thanks for reporting @Jethro87 - I probably should have updated the ticket. I've also found that this happens in non-tagging cases. I've spent quite a bit of time digging into this and trying to fix it.

The root cause is once again this Flutter rebuild policy that it won't let us configure in any way: https://github.com/flutter/flutter/issues/123305

Unfortunately, the Flutter team seems to have no interest in facilitating what I brought up in that ticket.

I have a test that reproduces the issue. I'm now looking for any way that I can hack Flutter to get rid of this.

If you haven't already, I would recommend giving a thumbs up on that ticket above. I don't expect the Flutter team to ever make that a priority, but all we can do is click the emoji and hope for the best.

**Comment by Jethro87** [comment]
Thanks @matthew-carroll. I have already upvoted the issue.

In the meantime, do you have insight on how I can avoid/debug this issue? In my app, I have my super_editor implementation, which I'm currently upgrading to catch up with all of the updates since atomic commands. I also have a basic super_editor implementation for testing, which nearly exactly follows `example_editor.dart`. 

This example editor does not throw these errors, while my implementation does. I've spent this morning going line by line to try and understand the difference that's causing this issue, but so far I've been unsuccessful. 

From your digging into this, can you point me in the right direction?

**Comment by matthew-carroll** [comment]
I can provide two different perspectives - one technical, one qualitative.

From a technical perspective, it looks like if an edit or some related change causes the full `SuperEditor` widget to rebuild, and also causes one of the overlays to rebuild, then it generates this error.

I've caused this bug in two ways on my end. 

The tagging system causes it because apparently I had the tagging system sending a change notification even when the tags didn't change. So I think it was scheduling a rebuild every time user the typed a character. Obviously, I shouldn't do that, so I'll only send the change notification when something changes. But even then, when the user actually inputs a tag, th error will come back.

The second place I've caused this is by adding support for changing the color of selected text. As the user drags a selection, I'm changing the attributions of the text. This is causing `SuperEditor` and the selection overlay to both invalidate on the same frame.

When both `SuperEditor` and an overlay invalidate, Flutter seems to be building the overlay before Flutter builds the document layout. Thus, the overlay tries to access text layout before there is any text layout. This is the issue that `ContentLayers` is supposed to solve for us, but again, due to Flutter's restriction on when things are built, I've been unable to hack `ContentLayers` to avoid this situation.
